pyodide-build/pyodide_build/pywasmcross.py

```diff
@@ -197,55 +203,199 @@ def replay_f2c(args, dryrun=False):
-    >>> replay_command(['gcc', 'test.c'], args, dryrun=True)
-    emcc test.c
-    ['emcc', 'test.c']
+
+def replay_genargs_handle_dashl(
+    arg: str, replace_libs: Dict[str, str], used_libs: Set[str]
+) -> Optional[str]:
     """
-    # some libraries have different names on wasm e.g. png16 = png
-    replace_libs = {}
-    for l in args.replace_libs.split(";"):
-        if len(l) > 0:
-            from_lib, to_lib = l.split("=")
-            replace_libs[from_lib] = to_lib
+    Figure out how to replace a `-lsomelib` argument.
 
-    # This is a special case to skip the compilation tests in numpy that aren't
-    # actually part of the build
-    for arg in line:
-        if r"/file.c" in arg or "_configtest" in arg:
-            return
-        if re.match(r"/tmp/.*/source\.[bco]+", arg):
-            return
-        if arg == "-print-multiarch":
-            return
-        if arg.startswith("/tmp"):
-            return
+    Parameters
+    ----------
+    arg
+        The argument we are replacing. Must start with `-l`.
 
-    if line[0] == "gfortran":
-        result = replay_f2c(line)
-        if result is None:
-            return
-        line = result
-        new_args = ["emcc"]
-    elif line[0] == "ar":
+    replace_libs
+        The dictionary of libraries we are replacing
+
+    used_libs
+        The libraries we've used so far in this command. emcc fails out if `-lsomelib`
+        occurs twice, so we have to track this.
+
+    Returns
+    -------
+        The new argument, or None to delete the argument.
+    """
+    assert arg.startswith("-l")
+    for lib_name in replace_libs.keys():
+        # this enables glob style **/* matching
+        if PurePosixPath(arg[2:]).match(lib_name):
+            arg = "-l" + replace_libs[lib_name]
+
+    if arg == "-lffi":
+        return None
+
+    # See https://github.com/emscripten-core/emscripten/issues/8650
+    if arg in ["-lfreetype", "-lz", "-lpng", "-lgfortran"]:
+        return None
+
+    # WASM link doesn't like libraries being included twice
+    # skip second one
+    if arg in used_libs:
+        return None
+    used_libs.add(arg)
+    return arg
+
+
+def replay_genargs_handle_dashI(arg: str, target_install_dir: str) -> Optional[str]:
+    """
+    Figure out how to replace a `-Iincludepath` argument.
+
+    Parameters
+    ----------
+    arg
+        The argument we are replacing. Must start with `-I`.
+
+    target_install_dir
+        The target_install_dir argument.
+
+    Returns
+    -------
+        The new argument, or None to delete the argument.
+    """
+    assert arg.startswith("-I")
+    if (
+        str(Path(arg[2:]).resolve()).startswith(sys.prefix + "/include/python")
+        and "site-packages" not in arg
+    ):
+        return arg.replace("-I" + sys.prefix, "-I" + target_install_dir)
+    # Don't include any system directories
+    if arg[2:].startswith("/usr"):
+        return None
+    return arg
+
+
+SYS_ROOT_RE = re.compile(",--sysroot=[^,]*")
```

Comment:
Are you sure the solution without the regex wasn't better? Also this will match anything that doesn't have "," including spaces and I don't think that's correct?

Reply:
Spaces shouldn't appear, unless it was something like `--sysroot="/a/path with spaces/"` in which case we should get rid of them. This does fail on `--sysroot="/a/path,with,commas/"`, but that should never happen I hope.

The previous solution without the regex was incorrect. However there is a better approach I can switch to.